content/en/blog/_posts/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm.md

```diff
@@ -0,0 +1,204 @@
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/github.jpg)
+
+Let's take a look at the parameters that we pass to Helm in the values file:
+
+* [**kubernetes/values.yaml**](https://github.com/kvaps/kubernetes-in-kubernetes/blob/master/deploy/helm/kubernetes/values.yaml)
+
+<img align="right" alt="Kubernetes is just a five binaries" style="height: 280px;" src="/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/5binaries.png">
+
+Beside **persistence** (storage parameters for the cluster), the Kubernetes control-plane components are described here: namely: **etcd-cluster**, **apiserver**, **controller-manager** and **scheduler**. These are pretty much standard Kubernetes components. There is a joke that «Kubernetes is just a five binaries». So here the configuration for these binaries is located.
+
+If you ever tried to bootstrap a cluster using kubeadm, then this config will remind you it's configuration. But in addition to Kubernetes entities, we also have an admin container. In fact, it is a container, which holds two binaries inside: **kubectl** and **kubeadm**. They are used to generate kubeconfig for the above components and to perform the initial configuration for the cluster. Also, in emergency case, you can always exec into it to check and manage your cluster.
+
+After the release [has been deployed](https://asciinema.org/a/407280), we will see a list of pods: **admin-container**, **apiserver** in two replicas, **controller-manager**, **etcd-cluster**, **scheduller** and the initial job that initializes our cluster. In the end we have a command, which allows you to get shell into the admin container, we can use it to see what is happening inside:
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot01.jpg)
+
+Also, let's take look at the certificates. If you've ever installed Kubernetes, then you know that it has a <irony>*scary* directory</irony> **/etc/kubernetes/pki** with a bunch of some certificates. In our case, we have fully automated management of them with cert-manager. Thus it is enough to pass all certificates paramaters to Helm during installation, and all the certificates will automatically be generated for our cluster.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot02.jpg)
+
+Looking at one of the certificates eg. apiserver, you can see that it has a list of DNS names and IP addresses. If you want to make this cluster accessible outside, then just describe the additional DNS names in the values file and update the release. This will update the certificate resource, and cert-manager will regeneraate the certificate. You'll no longer need to think about this. If kubeadm certificates need to be renewed at least once a year, here the cert-manager will take care and automatically renew them.
```

Comment:
```suggestion
Looking at one of the certificates, eg. apiserver, you can see that it has a list of DNS names and IP addresses. If you want to make this cluster accessible outside, then just describe the additional DNS names in the values file and update the release. This will update the certificate resource, and cert-manager will regenerate the certificate. You'll no longer need to think about this. If kubeadm certificates need to be renewed at least once a year, here the cert-manager will take care and automatically renew them.
```